### ISSUE TITLE
Inquiry Regarding the Project Workflow

### ISSUE THREAD

**Issue by qaz8788817** [NaughtyZZ/3D_facial_shape_attribute_translation_ssgmap#3]
Hi, after successfully compiling the '3D_face_dense_registration' executable, I returned to this repository to study how to use the project. Here is my plan: I will use the templates provided by 'registration' to register my current FaceScape raw data, eliminating problematic meshes. Afterward, I will clip the frontal part and finally write the data into the 'training_data' in hdf5 format to begin training. Is this workflow correct, or should I register directly using the '3D_template.ply'?

I apologize for the continuous inquiries these past weeks. Regarding the 'clipping' and 'training_data' part, could you provide some code for my reference? If it's inconvenient to upload here, my email is gn00088507@gmail.com.
Thank you again for your guidance; it has been immensely helpful to me.

**Comment by NaughtyZZ** [comment]
If you choose the TU model in Facescape, you can apply my registration program (in fact it corrects the grid routines and improve the original registration results) for all TU data in facescape. Then, the whole clipping process is to keep the provided indices of all (corrected ) TU models . The vertex indices of the frontal part are provided  in [./template/idx_Convert_TU_Models_FaceScape.mat](https://github.com/NaughtyZZ/3D_facial_shape_attribute_translation_ssgmap/blob/main/template/idx_Convert_TU_Models_FaceScape.mat). Finally, I found some of the data in TU models are not correct registered, while others are with wrong expression labels. You may check every data and eliminate those with problems. For these processes, my codes are fragmented and may not be useful to you. You can contact me if you have any problem in these processes.

**Comment by NaughtyZZ** [comment]
A better way to do this is to register to the raw data from scratch with the provide 3D template file instead of the "TU models". However, there are only 300+ raw data (800+ for all) in the facescape dataset when I download the data. The registration process from scratch also requires quite a lot of efforts to manually label some landmarks for large expressions. Although better registration provides better training data, it requires more efforts. Currently I only provide the model trained with corrected data from the original "TU models".

**Comment by qaz8788817** [comment]
Thank you very much for your response! 
I have some questions here. First, I would like to ask if it is still possible to use this model without performing the preprocessing of clipping the frontal part. The second question is, what is the structure of the default file inside "--data_path" in main.py?

Thank you.

**Comment by qaz8788817** [comment]
However, I am currently using Python to read registered PLY files and vertex index files. I can extract vertices from the PLY file like this:
![image](https://github.com/NaughtyZZ/3D_facial_shape_attribute_translation_ssgmap/assets/73918947/adee1940-4eb3-406c-a337-e17bd4cfcffb)
And I am facing difficulties in dealing with the faces. From the format, the values after the "face" keyword represent the indices of vertices. But after extracting the front-facing vertices, their order is different. Could you please advise on how to resolve the issue with the faces?

Here is my code to read ply file and filter vertex:
`
    def filter_ply_by_mat(ply_file, mat_file):
        # 讀取PLY文件
        ply_data = PlyData.read(ply_file)

        # 讀取MAT文件
        mat_data = h5py.File(mat_file, 'r')

        # 獲取MAT文件中的索引
        vertex_indices = mat_data['idxConvert'][:]  # 請替換為你實際的MAT文件中的索引變量名

        vertex_indices = vertex_indices.astype(int)

        # 根據MAT文件中的索引篩選PLY文件中的頂點和面
        filtered_vertices = []
        for i in np.nditer(vertex_indices):
            # print(i[1])
            i = i - 1
            if i < len(ply_data['vertex']):
                # 從PLY文件中取得對應的頂點
                vertex = ply_data['vertex'][int(i)]
                # 將頂點添加到列表
                filtered_vertices.append(vertex)

        # 創建新的PLY文件
        new_ply_data = PlyData([PlyElement.describe(np.array(filtered_vertices), 'vertex')])

        # 保存新的PLY文件
        new_ply_file = 'filtered_output.ply'
        new_ply_data.write(new_ply_file)

        print(f"Filtered PLY file saved as {new_ply_file}")
`

Thanks.

**Comment by NaughtyZZ** [comment]
1. Although you can extract the specific vertices for training, it would be better to perform automatic clipping by renewing the vertices and triangles（also for normals accordingly）. The number and order of the vertices are renewed, the triangle indices should also be renewed accordingly to form a mesh. In this way, it is more convinient to see the results for cross-validation. 
2. The file structure:

-----shape_data: NX3 (normalized shape in a roughly range [-1,1] with N vertices);

-----label_data: 25-bit codes, the first 20 one-hot codes for expressions, following 4 one-hot codes for gender and symmetry codes, and the last for age codes normalized to [-1,1]. The symmetry codes can be determined by expression labels, while others are provided by the FaceScape data. More details are in the paper.

-----Others: I stored them in the [grid.hdf5](https://github.com/NaughtyZZ/3D_facial_shape_attribute_translation_ssgmap/blob/main/grid.hdf5) file. They are precomputed weights for bi-directional differentiable sampling.

**Comment by qaz8788817** [comment]
I apologize; there may have been errors in the way I asked and expressed my question.

Based on your previous response, does it mean that the part of clipping the frontal part is included in the "preprocess input data" stage in the solver.py?

To be more direct, I would like to start using this project from scratch, including preprocessing the dataset (such as retaining only the face, marking labels and landmarks, etc.), and also generating the training_data.hdf5. Could you provide relevant code for me to reference?

I sincerely apologize for making such a request. As a newly enrolled master's student without senior peers in the laboratory researching the same field, I am currently interested in exploring the topic of predicting postoperative facial models using preoperative facial models in orthognathic surgery. Your paper serves as a source of inspiration for my research, providing me with a direction for further study and implementation.

Thank you for taking the time to answer my questions and guiding me on how to use it.

**Comment by NaughtyZZ** [comment]
You are welcomed. I will send some of these codes via email. The codes are fragmented and I hope they will be helpful to you.  Clipping the frontal part is not included in the solver.py. The clipping is simply retaining the provided vertices in [./template/idx_Convert_TU_Models_FaceScape.mat](https://github.com/NaughtyZZ/3D_facial_shape_attribute_translation_ssgmap/blob/main/template/idx_Convert_TU_Models_FaceScape.mat).

**Comment by qaz8788817** [comment]
> You are welcomed. I will send some of these codes via email. The codes are fragmented and I hope they will be helpful to you. Clipping the frontal part is not included in the solver.py. The clipping is simply retaining the provided vertices in [./template/idx_Convert_TU_Models_FaceScape.mat](https://github.com/NaughtyZZ/3D_facial_shape_attribute_translation_ssgmap/blob/main/template/idx_Convert_TU_Models_FaceScape.mat).

Thanks. Here’s my email: gn00088507@gmail.com
And I will work hard to learn how to execute this project from start to finish. thank you for your help. I hope you won't be bothered by my message. If I have any other questions, would you mind if I continue to ask you?
